test-suite/utilities.hpp

```diff
@@ -41,6 +41,27 @@
 #include <utility>
 #include <vector>
 
+// This adapts the BOOST_CHECK_SMALL and BOOST_CHECK_CLOSE macros to
+// support a struct as Real for arguments, while fully transparant to regular doubles.
+// Unfortunately boost does not provide a portable way to customize these macros' behaviour,
+// so we need to define wrapper macros QL_CHECK_SMALL etc.
+//
+// It is required to have a function `value` defined that returns the double-value
+// of the Real type (or a value function in the Real type's namespace for ADT).
+
+namespace QuantLib {
+    // overload this function in case Real is something different - it should alway return double
+    inline double value(double x) {
+        return x;
+    }
+}
+
+using QuantLib::value;
+
+#define QL_CHECK_SMALL(FPV, T)  BOOST_CHECK_SMALL(value(FPV), value(T))
+#define QL_CHECK_CLOSE(L, R, T) BOOST_CHECK_CLOSE(value(L), value(R), value(T))
```

Comment:
Is `Real` supposed to define an explicit conversion to double?  If so, this could be
```
#define QL_CHECK_SMALL(FPV, T)  BOOST_CHECK_SMALL(double(FPV), double(T))
#define QL_CHECK_CLOSE(L, R, T) BOOST_CHECK_CLOSE(double(L), double(R), double(T))
```

Reply:
A static cast to `double`, even if explicit, is dangerous to add to an AD framework, as tracking of derivatives would be lost. Most AD frameworks we are aware of don't allow that. 

Sometimes there are some `double(xxx)` or `static_cast<double>(xxx)` elsewhere in a large code-base. Such places would be undetected if an explicit conversion operation was added to `Real`, causing incorrect derivatives to be computed. So we believe having an overloadable `value` function is the better and more generic approach here.

Reply:
Understood, thanks.